explorer/tests/test_utils.py

```diff
@@ -17,46 +16,173 @@ class TestSqlBlacklist(TestCase):
+             ELSE 'Over $1000'
+          END
+        FROM Production.Product
+        ORDER BY ProductNumber ;
+        '''
+        passes, words = passes_blacklist(sql)
+        self.assertTrue(passes)
+
+    def test_select_with_subselect(self):
+        sql = '''SELECT a.studentid, a.name, b.total_marks
+            FROM student a, marks b
+            WHERE a.studentid = b.studentid AND b.total_marks >
+            (SELECT total_marks
+            FROM marks
+            WHERE studentid =  'V002');
+            '''
+        passes, words = passes_blacklist(sql)
+        self.assertTrue(passes)
+
+    def test_select_with_replace_function(self):
+        sql = "SELECT replace('test string', 'st', '**');"
+        passes, words = passes_blacklist(sql)
+        self.assertTrue(passes)
+
+    def test_dml_commit(self):
+        sql = "COMMIT TRANSACTION;"
+        passes, words = passes_blacklist(sql)
+        self.assertFalse(passes)
+
+    def test_dml_delete(self):
         sql = "'distraction'; deLeTe from table; " \
               "SELECT 1+1 AS TWO; drop view foo;"
         passes, words = passes_blacklist(sql)
         self.assertFalse(passes)
         self.assertEqual(len(words), 2)
-        self.assertEqual(words[0], 'DROP')
-        self.assertEqual(words[1], 'DELETE')
 
-    def test_queries_dropping_views_is_not_ok_and_not_case_sensitive(self):
-        sql = "SELECT 1+1 AS TWO; drop ViEw foo;"
-        self.assertFalse(passes_blacklist(sql)[0])
+    def test_dml_insert(self):
+        sql = "INSERT INTO products (product_no, name, price) VALUES (1, 'Cheese', 9.99);"
+        passes, words = passes_blacklist(sql)
+        self.assertFalse(passes)
 
-    def test_queries_containing_drop_in_word_is_ok(self):
-        sql = "SELECT * FROM student droptable WHERE name LIKE 'Robert%'"
-        self.assertTrue(passes_blacklist(sql)[0])
+    def test_dml_merge(self):
+        sql = '''MERGE INTO wines w
+            USING (VALUES('Chateau Lafite 2003', '24')) v
+            ON v.column1 = w.winename
+            WHEN NOT MATCHED
+              INSERT VALUES(v.column1, v.column2)
+            WHEN MATCHED
+              UPDATE SET stock = stock + v.column2;'''
+        passes, words = passes_blacklist(sql)
+        self.assertFalse(passes)
 
-    def test_sql_whitelist_ok(self):
-        app_settings.EXPLORER_SQL_WHITELIST = ['dropper']
-        sql = "SELECT 1+1 AS TWO; dropper ViEw foo;"
-        self.assertTrue(passes_blacklist(sql)[0])
+    def test_dml_replace(self):
+        sql = "REPLACE INTO test VALUES (1, 'Old', '2014-08-20 18:47:00');"
+        passes, words = passes_blacklist(sql)
+        self.assertFalse(passes)
+
+    def test_dml_rollback(self):
+        sql = "ROLLBACK TO SAVEPOINT my_savepoint;"
+        passes, words = passes_blacklist(sql)
+        self.assertFalse(passes)
+
+    def test_dml_set(self):
+        sql = "SET TIME ZONE 'PST8PDT';"
```

Comment:
@marksweb why forbid this? What's a better way to display results in a local timezone?

Reply:
Forbid this? Sorry I don't follow - I haven't forbidden this?

Reply:
This PR added `SET` to `app_settings.EXPLORER_SQL_BLACKLIST`, and the assert below seems to be checking that `SET TIME ZONE 'PST8PDT'` fails the blacklist. I noticed this because several queries I rely on started failing after an upgrade.

From a quick reading of https://www.postgresql.org/docs/current/sql-set.html, I don't notice ways in which it'd modify data in the database, but I'm not a SQL expert, and I haven't checked MySQL, etc.

Reply:
@covracer This PR just changed the order of the blacklisted items. `SET` was actually added about a year ago [here](https://github.com/groveco/django-sql-explorer/commit/ca31cc946de68bc7eb9a5b03601134a70e4b1564).

`SET` can be used to update values. See [here](https://www.w3schools.com/SQL/sql_ref_set.asp).

The blacklist check didn't work for some time. It may have been this PR which actually got it working correctly hence you noticing it.

Reply:
https://www.w3schools.com/SQL/sql_ref_set.asp discusses `SET` in the context of `UPDATE ... SET ... WHERE`. It seems to me like the entry for `UPDATE` would cover that.

`SET PASSWORD` for old versions of MySQL as described in #475 sounds like a great thing to block. Would it be possible to disallow `SET PASSWORD ...` while allowing `SET TIME ...`?